src/record/json-path.js

```diff
@@ -49,58 +33,87 @@ JsonPath.prototype.getValue = function() {
  * @param {Mixed} value
  *
  * @public
- * @returns {void}
+ * @returns updated value
  */
-JsonPath.prototype.setValue = function( value ) {
-	var node = this._record._$data,
-		i;
+module.exports.set = function( data, path, value, deepCopy ) {
+	var tokens = tokenize( path );
 
-	for( i = 0; i < this._tokens.length - 1; i++ ) {
-		if( node[ this._tokens[ i ] ] !== undefined ) {
-			node = node[ this._tokens[ i ] ];
+	if ( tokens.length === 0 ) {
+		return patch( data, value, deepCopy );
+	}
+
+	var oldValue = module.exports.get( data, path, false );
+	var newValue = patch( oldValue, value, deepCopy );
+
+	if ( newValue === oldValue ) {
+		return data;
+	}
+
+	var result = utils.shallowCopy( data );
+
+	var node = result;
+	for( var i = 0; i < tokens.length; i++ ) {
+		if ( i === tokens.length - 1) {
+			node[ tokens[ i ] ] = newValue;
 		}
-		else if( this._tokens[ i + 1 ] && !isNaN( this._tokens[ i + 1 ] ) ){
-			node = node[ this._tokens[ i ] ] = [];
+		else if( node[ tokens[ i ] ] !== undefined ) {
+			node = node[ tokens[ i ] ] = utils.shallowCopy( node[ tokens[ i ] ] );
+		}
+		else if( tokens[ i + 1 ] && !isNaN( tokens[ i + 1 ] ) ){
+			node = node[ tokens[ i ] ] = [];
 		}
 		else {
-			node = node[ this._tokens[ i ] ] = {};
+			node = node[ tokens[ i ] ] = Object.create( null );
 		}
 	}
 
-	node[ this._tokens[ i ] ] = value;
+	return result;
 };
 
+function patch( oldValue, newValue, deepCopy ) {
+	var i;
+
+	if ( utils.deepEquals( oldValue, newValue ) ) {
```

Comment:
I'm a little confused, won't this run recursively every time and outweigh the benefit of having the deepCopy flag? 

I can see it being an advantage when the flag is false since it will do literal object comparison, but when the flag is true it will have to do this on every object..


```
return deepCopy !== false ? utils.deepCopy( newValue ) : newValue;
```